### DISCUSSION TITLE
Optimistic UI + Single Source of Truth

### DISCUSSION THREAD

**Discussion by 124C41p** [leptos-rs/leptos#1197] [Q&A]
In the `counter_isomorphic` example (multi-user-mode), all clients see exactly the same counter, synchronized through SSE. In that way however, updating the count might be slow due to server latency. Suppose we want to improve the example by making the UI optimistic: Each time we send a count update request to the server, we immediately want to update our own counter. When we receive feedback through SSE, we want to update the counter again (the server being the single source of truth).

What is the proper way to do this in Leptos?

Some thoughts:

There are two relevant read signals for count in this case: One is given by the SSE, the other is given by the input signal of the server action (not exactly true in the example, but can be easily refactored in that way). A quick and dirty solution might be to create a third count signal and two effects such that each change to one of the former signals updates the third. But now we have a potential race condition: If the server communication is faster than the client updating itself for some reason, the client might override a recent server event with its outdated update request (violating the single source of truth principle).

A robust solution should be possible by enumerating all events, so that both signals contain the current count, and the corresponding event number. Now the third signal could just be a derived signal choosing the count value with higher event number (server signal wins on equality). However, this solution seems nontrivial and clumsy. Am I missing an elegant solution?

**Comment by gbj** [comment]
The designated way to do optimistic UI would be to use [`create_action`](https://docs.rs/leptos/latest/leptos/fn.create_action.html). An [`Action`](https://docs.rs/leptos/latest/leptos/struct.Action.html) has a `.input` method that returns a signal with the most recent value you've submitted, as well as a `.value` method that returns the most recent value. So the value could display `.input()` *or* the most recent from the server, whichever is fresher, so that a user's update would appear immediately.

This is very similar to what you're saying about the two relevant signals, in this case. I recognize the theoretical possibility of the race condition. However, I don't think it's actually possible; because of the synchronous and single-threaded nature of JavaScript/the browser's event loop, and the synchronous nature of the reactive system, I'm pretty sure it's not possible for an async event to arrive and be processed between the moment a signal updates and the moment that update is reflected in the DOM.

Either the event from the server arrives before the user clicks the button, in which case the user-button-clicked version wins, which is correct; or the event from the server arrives after the user clicks the button, in which case the server version wins, which is correct. I don't believe it's possible for the server-sent event to arrive and be processed between the button click and the DOM update.

**Reply by 124C41p** [reply]
Good point, I might actually be wrong about the race condition. But there is still something I am missing:

In the present situation we have two signals of type `ReadSignal<Option<i32>>`,  say `server_signal` and `action_input_signal`.  But what we actually want to bind to the UI is a derived signal `display_signal` which always shows the most recent value of both.

How do we define `display_signal`? Is there a _merge_ function in Leptos which does exactly what we need, or is this one of the rare occasions where you would actually use effects to update a signal (which is normally discouraged)?

**Reply by gbj** [reply]
In this specific case of a server-sent event, but with optimistic UI on your end, I'd actually rethink this a bit and say to use a single signal to hold the current value, which is set from both the loop waiting for updates from the stream, and the click handlers.

```rust
// maybe pseudocode
let (value, set_value) = create_signal(cx, None);
let mut source = EventSource::new(/* URL */).unwrap();
spawn_local(async move {
  while let Some(event) = source.next().await {
    set_value(/* whatever */);
  }
});

let on_click = move |_| {
  action.dispatch(());
  set_value(/* whatever */);
}
```

**Reply by 124C41p** [reply]
This causes a very unfortunate bug: Suppose a server call takes 5 seconds due to high latency. If the user quickly hits the "increase counter"-button 10 times, the counter gets immediately increased by 10. But after 5 seconds the server events arrive. Now the counter bounces back to the start value, and then gets increased again.
